FiniteStateMachine/src/statemachine/State.java

```diff
@@ -2,11 +2,9 @@
 
 import org.jetbrains.annotations.NotNull;
 
-import java.util.EnumSet;
+import java.util.Set;
 
 
-public interface State<E extends Enum<E> & State<E>> {
-    @NotNull EnumSet<E> nextStates();
```

Comment:
It's fine (and correct) to use `EnumSet` when actually creating the set, if the implementation of `State` is an enum. But `Gear.nextState` can't return an `EnumSet` because java's generics are all invariant, so you can't just change the return type.